src/libcadet/model/parts/ConvectionDispersionOperator.cpp

```diff
@@ -185,29 +186,21 @@ bool ConvectionDispersionOperatorBase::configure(UnitOpIdx unitOpIdx, IParameter
  */
 bool ConvectionDispersionOperatorBase::notifyDiscontinuousSectionTransition(double t, unsigned int secIdx)
 {
-	double prevVelocity = static_cast<double>(_curVelocity);
-
+	const int dirOld = _dir;
 	// If we don't have cross section area, velocity is given by parameter
-	if (_crossSection <= 0.0)
-		_curVelocity = getSectionDependentScalar(_velocity, secIdx);
-	else if (!_velocity.empty())
+	// We have both cross section area and interstitial flow rate
+	// _curVelocity has already been set to the network flow rate in setFlowRates()
+	// the direction of the flow (i.e., sign of _curVelocity) is given by _velocity
+if (_crossSection <= 0.0 || !_velocity.empty())
 	{
-		if (secIdx > 0)
-		{
-			const double dir = static_cast<double>(getSectionDependentScalar(_velocity, secIdx - 1));
-			if (dir < 0.0)
-				prevVelocity *= -1.0;
-		}
+		const double dirNew = static_cast<double>(getSectionDependentScalar(_velocity, secIdx));
+		_dir = (dirNew >= 0) ? 1 : -1;
 
-		// We have both cross section area and interstitial flow rate
-		// _curVelocity has already been set to the network flow rate in setFlowRates()
-		// the direction of the flow (i.e., sign of _curVelocity) is given by _velocity
-		const double dir = static_cast<double>(getSectionDependentScalar(_velocity, secIdx));
-		if (dir < 0.0)
-			_curVelocity *= -1.0;
+		if (dirOld * _dir < 0)
+			_curVelocity *= -1;
 	}
 
-	return (prevVelocity * static_cast<double>(_curVelocity) < 0.0);
+	return (dirOld * _dir < 0);
 }
```

Comment:
Okay, this got really convoluted. Let's have a look at the original logic:
```cpp
bool ConvectionDispersionOperatorBase::notifyDiscontinuousSectionTransition(double t, unsigned int secIdx)
{
	double prevVelocity = static_cast<double>(_curVelocity);

	// If we don't have cross section area, velocity is given by parameter
	if (_crossSection <= 0.0)
		_curVelocity = getSectionDependentScalar(_velocity, secIdx);
	else if (!_velocity.empty())
	{
		if (secIdx > 0)
		{
			const double dir = static_cast<double>(getSectionDependentScalar(_velocity, secIdx - 1));
			if (dir < 0.0)
				prevVelocity *= -1.0;
		}

		// We have both cross section area and interstitial flow rate
		// _curVelocity has already been set to the network flow rate in setFlowRates()
		// the direction of the flow (i.e., sign of _curVelocity) is given by _velocity
		const double dir = static_cast<double>(getSectionDependentScalar(_velocity, secIdx));
		if (dir < 0.0)
			_curVelocity *= -1.0;
	}

	return (prevVelocity * static_cast<double>(_curVelocity) < 0.0);
}

void ConvectionDispersionOperatorBase::setFlowRates(const active& in, const active& out, const active& colPorosity) CADET_NOEXCEPT
{
	// If we have cross section area, interstitial velocity is given by network flow rates
	if (_crossSection > 0.0)
		_curVelocity = in / (_crossSection * colPorosity);
}
```
We have three cases:
1. `_velocity` is non-empty and `_crossSection <= 0.0`: `_curVelocity` (including direction) is goverend by `_velocity`.
2. `_velocity` is non-empty and `_crossSection > 0.0`: The magnitude of `_curVelocity` is provided by the network in `setFlowRates()` and its direction is given by the sign of `_velocity`.
3. `_velocity` is empty and `_crossSection > 0.0`: Magnitude and direction are provided by the network in `setFlowRates()`.

First, `setFlowRates()` is called, then `notifyDiscontinuousSectionTransition()`. In case of dynamic flow rates, we have subsequent calls to `setFlowRates()`. These subsequent calls were resetting the direction of `_curVelocity`, which caused the bug.

In order to preserve the direction of the flow in subsequent calls, you've introduced the `_dir` variable and replaced `setFlowRates()`:
```cpp
void ConvectionDispersionOperatorBase::setFlowRates(const active& in, const active& out, const active& colPorosity) CADET_NOEXCEPT
{
	// If we have cross section area, interstitial velocity is given by network flow rates
	if (_crossSection > 0.0)
		_curVelocity = _dir * in / (_crossSection * colPorosity);
}
```
Now we need to handle the three cases via `notifyDiscontinuousSectionTransition()`:
```cpp
bool ConvectionDispersionOperatorBase::notifyDiscontinuousSectionTransition(double t, unsigned int secIdx)
{
	// setFlowRates() was called before, so _curVelocity has direction dirOld
	const int dirOld = _dir;

	if (_crossSection <= 0.0)
	{
		// Use the provided _velocity (direction is also set), only update _dir
		_curVelocity = getSectionDependentScalar(_velocity, secIdx);
		_dir = (_curVelocity >= 0.0) ? 1 : -1;
	}
	else if (!_velocity.empty())
	{
		// Use network flow rate but take direction from _velocity
		_dir = (getSectionDependentScalar(_velocity, secIdx) >= 0.0) ? 1 : -1;

		// _curVelocity has correct magnitude but previous direction, so flip it if necessary
		if (dirOld * _dir < 0)
			_curVelocity *= -1.0;
	}

	// Remaining case: _velocity is empty and _crossSection <= 0.0
	// _curVelocity is goverend by network flow rate provided in setFlowRates().
	// Direction never changes (always forward, that is, _dir = 1)-
	// No action required.

	// Detect change in flow direction
	return (dirOld * _dir < 0);
}
```

Reply:
Ah, I get it now! I did not fully understand how flow rates work if no cross section area is provided.